### PR TITLE
Get rid of the write progress callbacks

### PR DESCRIPTION
**Pull request by rklaehn**
instead have a wrapper for an AsyncSliceWriter that can optionally produce progress events.

As requested in some long gone PR review

### CODE REVIEW

**Comment by dignifiedquire** [comment]
Something is strange about these commits

**Comment by rklaehn** [comment]
> Something is strange about these commits

Yeah, you can ignore the commit history. It contains stuff that has already been squash merged. Just look at the diff.

File: src/tokio_util.rs [hunk 1]
```diff
@@ -55,6 +57,82 @@ where
...
+    }
+
+    /// Return the inner writer
+    pub fn into_inner(self) -> W {
+        self.0
+    }
+}
+
+impl<W: AsyncSliceWriter + Send + 'static> AsyncSliceWriter for ProgressSliceWriter<W> {
+    type WriteFuture = BoxFuture<'static, (Self, io::Result<()>)>;
```
[the remark this answers is not in the document]
Comment (rklaehn) [reply src/tokio_util.rs:113]: Unfortunately not. These things are written in state passing style, so they return Self. So you either have to write a custom future, or box. Here I was lazy so I boxed.

The state passing style is the only thing I have found to _in principle_ allow io for Send things with unboxed futures without having to go down all the way to poll fns, but you still have to manually write those futures.

Don't put your hope on async traits - they are not going to be `Send`. If you need then to be `Send`, the official recommendation is to continue to use the async_trait crate which of course boxes.

https://blog.rust-lang.org/inside-rust/2022/11/17/async-fn-in-trait-nightly.html#limitation-spawning-from-generics


So - I think the best medium term remedy for all this insanity would be to just not require `Send` for our IO things. Then we could define the io primitives in reference style which matches more how sync IO works. Of course that would mean going to a thread per core architecture where just some kind of abstract file handle is Send, but an open file is not.

Then you could do something like this:

```rust
trait AsyncSliceReader {
  type ReadFuture<a'> = Future<Output = io::Result<()>> + 'a where 'a: Self;
  fn read_at<'a>(&'a mut self, offset: u64, data: &'a mut [u8]) -> ReadFuture<'a>
}
```

(I tried this exact thing, but it does not work with Send due to "higher kinded lifetime yadda yadda yadda")

[the remark this answers is not in the document]
Comment (dignifiedquire) [reply src/tokio_util.rs:113]: wow, that is a very sad state of the async world.. 😭

**Review by dignifiedquire — APPROVED** [review]

File: src/tokio_util.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (rklaehn) [reply src/tokio_util.rs:113]: I think state passing style is not that bad, except that you would probably need some kind of new loop construct or macro to make it pleasant to use.

State passing style is without alternative anyway for FSMs where a state transition can change the type of the state.

**Pull request merged**